### PR TITLE
feat(notifications): add notifications package with providers. 

### PR DESCRIPTION
**Pull request by maneesha-xyz**
## Summary

Breaking #390 into smaller PR's.

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] Performance improvement
- [ ] Other (please describe):

## Changes Made

- List the main changes made in this PR
- Be specific about what was modified, added, or removed
- Include any architectural changes

## Testing

- [ ] I have tested this change locally
- [x] I have added unit tests for new functionality
- [ ] I have tested edge cases
- [x] All existing tests pass

## Checklist

- [x] My code follows the project's style guidelines
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have checked that there are no merge conflicts

## Related Issues

#390 

## Screenshots/Demo

N/A

## Additional Notes

N/A

## Deployment Notes

N/A

### CODE REVIEW

File: backend/pkg/notifications/loader.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	"os"
+)
+
+// expandEnv expands ${VAR_NAME} placeholders in data with their environment variable values.
+// Values are JSON-escaped before substitution so special characters cannot corrupt the JSON document.
+// Returns an error listing all unset variables if any are missing.
+func expandEnv(data []byte) ([]byte, error) {
+	var missing []string
+	expanded := os.Expand(string(data), func(key string) string {
+		val, ok := os.LookupEnv(key)
```
Comment (ginaxu1) [inline backend/pkg/notifications/loader.go]: antipattern to do string substitution on raw JSON before unmarshaling

better Go practice is to unmarshal the JSON into a Go struct or `map[string]any` first, and then do the `os.ExpandEnv`; or you can use existing libraries that handles env variable overriding natively and safely

**Review by ginaxu1 — CHANGES_REQUESTED** [review]

File: backend/pkg/notifications/types.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+type Request struct {
+	Channel  ChannelType
+	To       string
+	Subject  string
+	Body     string
+	HTMLBody string
+}
+
+func (r Request) Validate() error {
```
Comment (ginaxu1) [inline backend/pkg/notifications/types.go]: you may want to use pointer receiver `func (r *Request) Validate() error {`, because a value receiver means the entire struct is copied every time Validate() is called

File: backend/pkg/notifications/loader.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+		if !ok {
```
Comment (ginaxu1) [inline backend/pkg/notifications/loader.go]: better check if missing variable was already added, otherwise you get duplicate missing vars 
```
if !ok {
			if !slices.Contains(missing, key) {
				missing = append(missing, key)
			}
			return "${" + key + "}"
		}
```

**Review by ginaxu1 — CHANGES_REQUESTED** [review]

File: backend/pkg/notifications/types.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+	switch r.Channel {
+	case ChannelSMS, ChannelEmail:
+		// valid
+	default:
+		return fmt.Errorf("invalid or missing channel: %q", r.Channel)
```
Comment (ginaxu1) [inline backend/pkg/notifications/types.go]: lines 27, 30, 33 

antipattern to use `fmt.Errorf` for static error strings, because it prevents callers from using errors.Is() to check for specific validation failures.

better do like this:
```go
var (
	ErrInvalidChannel = errors.New("invalid or missing channel")
	ErrToRequired     = errors.New("to address is required")
	ErrBodyRequired   = errors.New("body or html_body is required")
)

func (r *Request) Validate() error {
	switch r.Channel {
	case ChannelSMS, ChannelEmail:
		// valid
	default:
		return fmt.Errorf("%w: %q", ErrInvalidChannel, r.Channel)
	}
	// ...
}
```

**Review by ginaxu1 — CHANGES_REQUESTED** [review]

File: backend/pkg/notifications/config.go [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+import "fmt"
+
+type Config struct {
+	Path string
+}
+
+func (c Config) Validate() error {
+	if c.Path == "" {
+		return fmt.Errorf("NOTIFICATION_CONFIG_PATH is required")
```
Comment (ginaxu1) [inline backend/pkg/notifications/config.go]: no need to hardcode `NOTIFICATION_CONFIG_PATH`.

`Config` struct doesn't need to know how the field's populated. 

better return a generic error, using `errors.New` instead of fmt.Errorf when there is no string formatting happening: `return errors.New("notification config path is required")`

**Review by ginaxu1 — CHANGES_REQUESTED** [review]

File: backend/pkg/notifications/loader.go (hunk 1 above)
Comment (maneesha-xyz) [reply backend/pkg/notifications/loader.go]: Since expansion only happens inside quoted JSON strings and escaping is handled safely, a recursive unmarshaller feels like unnecessary complexity for the current use case.

File: backend/pkg/notifications/types.go [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+	ChannelEmail ChannelType = "email"
+)
+
+var (
+	ErrInvalidChannel = errors.New("invalid or missing channel")
+	ErrToRequired     = errors.New("to address is required")
+	ErrBodyRequired   = errors.New("body or html_body is required")
+)
+
+type Request struct {
```
Comment (ginaxu1) [inline backend/pkg/notifications/types.go:21]: not convinced this is best approach: current `Request` struct forces all channels to share the exact same payload format. SMS doesn't need have Subject or an HTMLBody

@maneeshaxyz what do you think about
```
type EmailPayload struct {
	To       string
	Subject  string
	Body     string
	HTMLBody string
}

type SMSPayload struct {
	To   string
	Body string
}

// Update the interface to accept a generic payload or specific methods
type Provider interface {
	Type() ChannelType
	Configure(cfg json.RawMessage) error
	Send(ctx context.Context, payload any) error // implementation type-asserts the payload
}
```

**Review by ginaxu1 — CHANGES_REQUESTED** [review]

File: backend/pkg/notifications/types.go (hunk 6 above)
Comment (ginaxu1) [reply backend/pkg/notifications/types.go:21]: always make sure you add struct tags (because relying on Go's implicit field matching is brittle)

```
Channel  ChannelType `json:"channel"`
To       string      `json:"to"`
Subject  string      `json:"subject,omitempty"`
Body     string      `json:"body"`
HTMLBody string      `json:"html_body,omitempty"`
```

File: backend/pkg/notifications/loader.go [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+)
+
+// expandEnv expands ${VAR_NAME} placeholders in data with their environment variable values.
+// Values are JSON-escaped before substitution so special characters cannot corrupt the JSON document.
+// Returns an error listing all unset variables if any are missing.
+func expandEnv(data []byte) ([]byte, error) {
```
Comment (ginaxu1) [inline backend/pkg/notifications/loader.go]: on 2nd thought why can't we use a standard configuration library like viper or envconfig to safely handle this?

**Review by ginaxu1 — CHANGES_REQUESTED** [review]

**Review by sthanikan2000 — DISMISSED** [review]
lgtm

File: backend/pkg/notifications/types.go (hunk 6 above)
Comment (sthanikan2000) [reply backend/pkg/notifications/types.go:21]: @ginaxu1 Let's not worry about this now. 
`type.go` should have Request struct only and it should be like follows:
```go
type Request struct {
    TemplateCode string
	UserID string
	Channels []ChannelType
	Data map[string]any
}
```

CC: @mushrafmim , @maneeshaxyz

**Review by sthanikan2000 — APPROVED** [review]
lgtm

**Review by ginaxu1 — APPROVED** [review]
